upload large files doc (#1369)
VNG-Realisatie/gemma-zaken: docs/_content/ontwikkelaars/handleidingen-en-tutorials/large-files.md

---
title: "Large files upload tutorial"
weight: 100
---

In this tutorial we explore the ways to upload files to DRC depending on the file size. 

The most relevant component for this tutorial:

> * DRC: for creating documents and uploading files

Reviewer: Might be nitpicking, but you could maybe add a link to the openAPI schema of the DRC (https://documenten-api.vng.cloud/api/v1/schema/) here as a reference

Author: The link to DRC OAS is added

## The requirements for this tutorial

> * `docker` and `docker-compose` to host the components locally on your (development machine). 

Reviewer: ```suggestion
* `docker` and `docker-compose` to host the components locally on your development machine. 
```

>   See [installatie en configuratie](./installatie-en-configuratie) for a detailed description.

Reviewer: I'm not sure if this is because the PR is not merged yet, but the link doesn't work for me

Author: The internal links in the whole "handleidingen-en-tutorials" folder don't work for me in Github. But they work using Jekyll, and I supposed that was a point. So I made these links consistent with others. Do they work with Jekyll for you?

> * The [eenmalige setup](./eenmalige-setup) has been carried out.

Reviewer: Same applies to this link

Author: please take a look at the comment above

## Overview of file upload process in DRC

The main resource of DRC API is `EnkelvoudigInformatieObject` which supports creating and editing documents. 

> It's content can be divided into metadata (`identificatie`, `creatiedatum`, `titel` and etc) and file content.

Reviewer: ```suggestion
Its content can be divided into metadata (`identificatie`, `creatiedatum`, `titel` and etc) and file content.
```

The main metadata of file content used by DRC is file size derived from `bestandsomvang` attribute. 

> The process of uploading files is chosen based on the comparision of the value of `bestandsomvang` and the maximum file size.

Reviewer: ```suggestion
The process of uploading files is chosen based on the comparison of the value of `bestandsomvang` and the maximum file size.
```

There can be 3 options:

> * file size = 0, i.e. document contains only metadata without file content. Document is created during 1 request to DRC.

Reviewer: I would probably stick to `EnkelvoudigInformatieObject` (or InformatieObject) instead of using Document, as using both interchangeably might be confusing to some

> * file size <= maximum size. In this case file content is expected to be in `inhoud` attribute in 1 request.

Reviewer: ```suggestion
* file size <= maximum size. In this case file content is expected to be in `inhoud` attribute and the InformatieObject is created in 1 request.
```

> * file size > maximum size. In this case file content should be divided by parts and each part should be upload in the separate request.

Reviewer: ```suggestion
* file size > maximum size. In this case file content should be divided by parts and each part should be uploaded in a separate request.
```

## To work

> In this tutorial we will consider all three ways to create document base on the files size.

Reviewer 2: Some grammar improvements:
In this tutorial we will consider all three ways to create a document based on the file size.

> ### Configure Maximujm file size

Reviewer: ```suggestion
### Configure Maximum file size
```

The process choice depends on maximum file size. 
The default value of this parameter = 4GB (or 4294967296 Bytes)

### Configure ZTC

> Creation documents require the link to `informatieobjecttype`. 

Reviewer: ```suggestion
Creating documents requires the link to `informatieobjecttype`. 
```

          "zaaktypen": [],
          "besluittypen": [],
          "informatieobjecttypen": [
              "<informatieobjecttype url>" 
          ]
      }
   ]
   ```

Now we have `informatieobjecttype url` which we will use in the next step

### Create document

Now we will create document with 3 different ways to upload files.  

#### Create document without file. 

To create a document only with metadata we need:

* leave `inhoud` field empty
* `bestandsomvang` = 0 

    Example request:

    ```http
    POST http://<drc-ip>:8000/api/v1/enkelvoudiginformatieobjecten HTTP/1.0
    Authorization: Bearer abcd1234
    Content-Type: application/json
   
    {
        "identificatie": "12345",
        "bronorganisatie": "123456782",
        "creatiedatum": "2019-06-27",
        "titel": "detailed summary",
        "auteur": "document auteur",
        "taal": "eng",
        "bestandsomvang": 0,
        "bestandsnaam": "file_name",
        "informatieobjecttype": "<informatieobjecttype url>",
        "vertrouwelijkheidaanduiding": "openbaar"
    }
    ``` 
    
The response contains the created document without lock. 

#### Create document with small file

> To create the document with file size less or equal maximum file size we need:

Reviewer: ```suggestion
To create the document with file size less than or equal to maximum file size we need:
```

    The response contains this lock key and `bestandsdelen` part with information for file parts uploading. 
    
    Part of example response:

    ```json
    {
        "url": "http://<drc-ip>:8000/api/v1/enkelvoudiginformatieobjecten/<uuid>",
        ...
        "bestandsomvang": 5000000000,
        "locked": true,
        "bestandsdelen": [
            {
                "url": "http://<drc-ip>:8000/api/v1/bestandsdelen/<uuid1>",
                "volgnummer": 1,
                "omvang": 4294967296,
                "voltooid": false
            },
            {
                "url": "http://<drc-ip>:8000/api/v1/bestandsdelen/<uuid2>",
                "volgnummer": 2,
                "omvang": 705032704,
                "voltooid": false
            }
        ],
        "lock": "abcd"
    }
    ```

    Each of `bestandsdelen` object includes information necessary for upload of file parts:
    
    * `url`: the link for uploading the file part
    * `volgnummer`: the sequence number of this part. The order of merging file parts after their upload depends on this attribute.
    * `omvang`: the size of the file part in Bytes

>     * `voltooid`: status it this file part is uploaded yet. 

Reviewer: ```suggestion
    * `voltooid`: boolean that indicates if this file part is uploaded or not. 
```

2. Upload file parts

    Now we need to split our file into file parts such as their size would match the sizes of `bestandsdelen` objects.
    

>     After splitting the file we need to upload each file part in the separate request

Reviewer: ```suggestion
    After splitting the file we need to upload each file part in a separate request
```

    
    In the result `voltooid` should be changed to "true".

3. Unlock the document

    After uploading all the parts it is time to unlock the document. 
    By unlocking the user shows that the uploading process is finished and the file can be gathered together from uploaded file parts.

    ```http
    POST http://<drc-ip>:8000/api/v1/enkelvoudiginformatieobjecten/<uuid>/unlock HTTP/1.0
    Authorization: Bearer abcd1234
    Content-Type: application/json

    {
        "lock": "abcd"
    }
    ```

    During unlock all file parts were merged into the whole file, all temporary file parts were removed.
    
4. Request the document

    We can request the created document to see the changes.

    ```http
    GET http://<drc-ip>:8000/api/v1/enkelvoudiginformatieobjecten/<uuid> HTTP/1.0
    Authorization: Bearer abcd1234
    Content-Type: application/json
    ```

    Part of the example response:

    ```json
    {
        "url": "http://<drc-ip>:8000/api/v1/enkelvoudiginformatieobjecten/<uuid>",
        ...
        "inhoud": "http://<drc-ip>:8000/api/v1/enkelvoudiginformatieobjecten/<uuid>/download?versie=1",
        "locked": false,
        "bestandsdelen": []
    }
    ```

>     Now the document is unlocked, all `bestandsdelen` are removed and the file content can be download via the `inhoud` url

Reviewer: ```suggestion
    Now the document is unlocked, all `bestandsdelen` are removed and the file content can be downloaded via the `inhoud` url
```

> ###Summary

Reviewer: ```suggestion
### Summary
```

In this tutorial in ZTC we:

* created Catalogus;
* created Informatieobjecttype 

> After that we explore 3 options to create a document in DRC:

Reviewer: ```suggestion
After that we have explored 3 options to create a document in DRC:
```

* without a file

> * with the small file

Reviewer: ```suggestion
* with a small file
```

> * with the large file (multiple step approach)

Reviewer: ```suggestion
* with a large file (multiple step approach)
```
